lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,708 @@
+        # value of pcbBytesNeeded
+        FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+          # return value will be false from this call, since it's designed
+          # to fail. Just ignore it
+          QueryServiceStatusEx(
+            service,
+            SC_STATUS_PROCESS_INFO,
+            nil,
+            0,
+            bytes_pointer
+          )
+          size_required = bytes_pointer.read(:dword)
+          FFI::MemoryPointer.new(size_required) do |ssp_ptr|
+            status = SERVICE_STATUS_PROCESS.new(ssp_ptr)
+            success = QueryServiceStatusEx(
+              service,
+              SC_STATUS_PROCESS_INFO,
+              ssp_ptr,
+              size_required,
+              bytes_pointer
+            )
+            if success == FFI::WIN32_FALSE
+              raise Puppet::Util::Windows::Error.new(_("Service query failed"))
+            end
+          end
+        end
+        status
+      end
+      private :query_status
+
+      # @api private
+      # perform QueryServiceConfigW on a windows service and return the
+      # result
+      #
+      # @param [:handle] service handle of the service to query
+      # @return [QUERY_SERVICE_CONFIGW struct] the result of the query
+      def query_config(service)
+        config = nil
+        size_required = nil
+        # Fetch the bytes of memory required to be allocated
+        # for QueryServiceConfigW to return succesfully. This
+        # is done by sending NULL and 0 for the pointer and size
+        # respectively, letting the command fail, then reading the
+        # value of pcbBytesNeeded
+        FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+          # return value will be false from this call, since it's designed
+          # to fail. Just ignore it
+          QueryServiceConfigW(service, nil, 0, bytes_pointer)
+          size_required = bytes_pointer.read(:dword)
+          FFI::MemoryPointer.new(size_required) do |ssp_ptr|
+            config = QUERY_SERVICE_CONFIGW.new(ssp_ptr)
+            success = QueryServiceConfigW(
+              service,
+              ssp_ptr,
+              size_required,
+              bytes_pointer
+            )
+            if success == FFI::WIN32_FALSE
+              raise Puppet::Util::Windows::Error.new(_("Service query failed"))
+            end
+          end
+        end
+        config
+      end
+      private :query_config
+
+      # @api private
+      # waits for a windows service to report final_state if it
+      # is in pending_state
+      #
+      # @param [:handle] service handle to the service to wait on
+      # @param [Integer] pending_state the state to wait on
+      # @param [Integer] final_state the state indicating the transition is finished
+      # @return [bool] 'true' once the service is reporting final_state,
+      #   'false' if the service was not in pending_state or finaL_state
+      def wait_for_pending_transition(service, pending_state, final_state)
```

Comment:
Can / should this be user configurable, rather than relying on `:dwWaitHint`?

Reply:
That is a good question. I was actually thinking about this, and I'm not sure which path to take with this scenario for puppet 5.5.x

I think in puppet 6 we should probably make it user configurable (at least optionally), but I'm unsure if that is too dangerous a change for a Z release (and thus we should keep it out of 5.5.x)

Since we were using net.exe before there was no wait on pending anyways, and using the wait hint is better (IMO) than just trying to guess what a good time is.  This is the recommended way MS describes in their examples on how to wait for a pending transition.

But the question does remain if we think it would be safe enough to land what is essentially a small new feature in a Z release or if it might effect the stability of the product.

What do you think @e?

Reply:
I agree with you that it's safe as-is. Maybe file a ticket on this one.

I just wasn't sure (not having read the docs yet) if it's possible to get stuck here infinitely waiting on an unresponsive service.

Reply:
That shouldn't be possible, I can look in to a test for this but the logic is built in such a way that unless the service reports an increment to dwCheckPoint puppet fails after the wait hint elapses, so unless someone builds their service logic to infinitely increment the checkpoint it should always fail at some point.

I might file two tickets here: 1. add a new feature to puppet 6 to allow for configurable timeout, and 2. add a base timeout that stops the logic after some long period no matter what ( I think, for example, the scm will bail on a service after something like two minutes regardless of wait hint or check point)

Reply:
Ah, I forgot about that part: yeah MS themselves in their examples suggest this: the master minimum wait time is at least 1 second, and the master maximum wait time is 10 seconds. Here is the example I'm referring to: https://docs.microsoft.com/en-us/windows/desktop/Services/starting-a-service